Try: Define the minimum boilerplate code for a block-based theme

kjellr opened the pull request:
There's just a little bit of boilerplate necessary to build a block-based theme today. This PR tries to collect that all in one place.

```
assets
    alignments-front.css
block-templates
    index.html
experimental-theme.json
functions.php
index.php
style.css
```

Theoretically you could use this today as a base to build a theme from, but I gathered it mostly to start identifying the remaining pieces that (hopefully) Gutenberg can start handling automatically. I'll leave some code comments below. 

Some of the `add_theme_support` declarations and the (mostly) empty `experimental-theme.json` file could technically be removed, but I kept them in for discussion purposes. 

cc @jorgefilipecosta and @nosolosw for your thoughts.

kjellr commented on blockbase/assets/alignments-editor.css:
```diff
@@ -0,0 +1,15 @@
+/*
```
This file can be removed after https://github.com/WordPress/gutenberg/pull/26853 is merged.

kjellr commented on blockbase/assets/alignments-front.css:
```diff
@@ -0,0 +1,79 @@
+/*
```
This file would hopefully be removed once Gutenberg provides front-end alignments. https://github.com/WordPress/gutenberg/issues/26633

kjellr commented on blockbase/experimental-theme.json:
```diff
@@ -0,0 +1,28 @@
+					"units": [ "px", "em", "rem", "vh", "vw" ]
+				}
+			},
+			"custom": {
```
These rules are only used in the alignments stylesheets above, so if those are removed, these could be removed too.

kjellr commented on blockbase/functions.php:
```diff
@@ -0,0 +1,37 @@
+	function blockbase_support()  {
+
+		// Alignwide and alignfull classes in the block editor
+		add_theme_support( 'align-wide' );
```
Should block-based themes opt-into wide/full alignments by default?

kjellr commented on blockbase/functions.php:
```diff
@@ -0,0 +1,37 @@
+		add_theme_support( 'align-wide' );
+
+		// Adding support for core block visual styles.
+		add_theme_support( 'wp-block-styles' );
```
Should block-based themes opt-into the opinionated styles by default? (If not, this should be available as a theme.json setting)

kjellr commented on blockbase/functions.php:
```diff
@@ -0,0 +1,37 @@
+<?php
+
+if ( ! function_exists( 'blockbase_support' ) ) :
+	function blockbase_support()  {
```
Many of these settings should either be activated by default for block-based themes, or available as `theme.json` options:

```
add_theme_support( 'align-wide' );
add_theme_support( 'wp-block-styles' );
add_theme_support( 'responsive-embeds' );
add_theme_support( 'experimental-link-color' );
```

scruffian replied:
Is there an issue for this in Gutenberg?

kjellr replied:
I just opened https://github.com/WordPress/gutenberg/issues/26901 to discuss.

kjellr commented on blockbase/functions.php:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Enqueue scripts and styles.
+ */
+function blockbase_scripts() {
```
Just thinking out loud... if we can eliminate the add_theme_support calls above, this file would only exist to enqueue the stylesheets. Would it be possible to enqueue the stylesheets from within `theme.json` instead?

kjellr replied:
I've opened an issue here to discuss: https://github.com/WordPress/gutenberg/issues/26902

jorgefilipecosta commented on emptytheme/experimental-theme.json:
```diff
@@ -0,0 +1,37 @@
+				"customLineHeight": true,
+				"fontFamilies": [ ],
+				"fontSizes": [ ],
+				"fontStyles": [ ],
```
I guess in most cases one should not have to define fontStyles, fontWeights, textDecorations, and textTransforms. I guess the future one may define per font family. These API's are very new I would probably not include them yet as they will probably change.

kjellr replied:
Yeah, that all makes sense. I've removed for now.

jorgefilipecosta commented on emptytheme/functions.php:
```diff
@@ -0,0 +1,38 @@
+		add_theme_support( 'responsive-embeds' );
+
+		// Add support for experimental link color control.
+		add_theme_support( 'experimental-link-color' );
```
I guess we should enable link color using the theme.json. global.settings.color.link = true;,

jorgefilipecosta replied:
Hi @kjellr, did you faced any problem enabling link color using theme.json global.settings.color.link = true;?

kjellr replied:
Good call! I just added this in c75159f and it works fine. 👍

jorgefilipecosta commented on emptytheme/style.css:
```diff
@@ -0,0 +1,15 @@
+/*
+Theme Name: Empty Theme
```
Should we allow all this header information to be specified in theme.json? @aristath, @mtias, @nosolosw

It seems strange that we force a theme to have a CSS file. I guess a theme with just theme.json should be possible.

scruffian replied:
I was thinking the same. That would be nice.

pattonwebz replied:
The theme directory (and core as well) would need, what I assume, to be a pretty sizable overhaul to do away with the need to have any stylesheet.

aristath replied:
Yeah, currently core detects themes from their style.css file. So if we want to allow these to be inside the `theme.json` file, a patch would have to be submitted to core at the same time so that these can be properly used.
I agree though, the stylesheet will in many cases be just a just file used to define the theme.

jorgefilipecosta replied:
I guess as theme.json involves we should allow just a theme.json theme. But this change would touch many parts of our code and is not something essential. Also as we are I guess most themes will still need at least some lines of CSS code. But once theme.json expands having theme information there seems like a good path.

MaggieCabrera commented:
We've been testing this PR this morning and found that featured images are not enabled by default 😃 

We could add `add_theme_support( 'post-thumbnails' );` to `functions.php`

The docs say:

> To retain backward compatibility, add_theme_support declarations are retrofit in the proper categories. 
> If a theme uses add_theme_support('disable-custom-colors'), it’ll be the same as set global.settings.color.custom to false. If the experimental-theme.json contains any settings, these will take precedence over the values declared via add_theme_support.

But I tried to set global.settings.post.thumbnails = true on the json and it did nothing. I probably did it wrong.

kjellr commented:
Thanks! I don't think this option exists yet for `theme.json`, so I've added it to the list in https://github.com/WordPress/gutenberg/issues/26901. 

For now, I've added the traditional `add_theme_support( 'post-thumbnails' );` to `functions.php`.

MaggieCabrera commented on emptytheme/assets/alignments-front.css:
```diff
@@ -0,0 +1,79 @@
+.wp-site-blocks .alignfull {
+	transform: translateX(calc(0px - var(--wp--custom--width-horizontal)));
+	width: calc(100% + (2 * var(--wp--custom--width-horizontal)));
+	max-width: calc(100% + (2 * var(--wp--custom--width-horizontal)));
```
Where is this defined? Should it be `--wp--custom--margin-horizontal`?

Firefox devtools tells me it is not set:

<img width="591" alt="Screenshot 2020-11-18 at 13 27 46" src="https://user-images.githubusercontent.com/3593343/99530869-2bba4300-29a2-11eb-9746-41e3301945c0.png">

nosolosw replied:
Those are generated from the theme.json declaration (see docs for [free-from custom properties](https://developer.wordpress.org/block-editor/developers/themes/theme-json/#free-form-css-custom-properties)). From what I see in [the file](https://github.com/WordPress/theme-experiments/pull/81/files#diff-fbd7c92cbefc0f9735d61095ae8e3b5a5b42521a1853477566cd54d981d7ba1aR16), the ones that will be generated are: `--wp--custom--width--default`, `--wp--custom--width--wide`, and `--wp--custom--margin--horizontal`.

kjellr replied:
Good catch! That should be `--wp--custom--margin--horizontal`. I've updated that in 5a32027

jorgefilipecosta reviewed (commented):
Should we also include a simple homepage template that lists the posts as part of the minimum boilerplate?

kjellr commented:
> Should we also include a simple homepage template that lists the posts as part of the minimum boilerplate?

Yeah, we could — We'd technically need both an `index.html` and a `singular.html` template to make it all work right now. And since it's more than one template, then a `header.html` template part makes sense, to house the site title. 😄 It starts snowballing from there. 

Without any templates, the theme doesn't technically work by default. So it probably makes sense to add in that baseline at least. I've added them in 2c6d4e1.

primenest reviewed (commented):
what would the `emptytheme/assets/block-template-parts` file be used for?

kjellr commented:
> what would the emptytheme/assets/block-template-parts file be used for?

As of [the last commit](2c6d4e1), that folder contains a reusable [header template part](https://github.com/WordPress/theme-experiments/blob/2c6d4e15f35093f0531259c58471ab06e1d7fe79/emptytheme/block-template-parts/header.html). This is used at the top of the `index.html` and `singular.html` templates.

mintplugins commented:
@kjellr Do you know if the plans to make the FSE HTML files into PHP files should be reflected here yet?

kjellr commented:
> Do you know if the plans to make the FSE HTML files into PHP files should be reflected here yet?

Not yet — this PR reflects the current state, not a planned future state.

jffng reviewed (commented):
`emptytheme/assets/block-template-parts` is an empty file that I think was accidentally created and checked in?

jorgefilipecosta commented on emptytheme/block-templates/index.html:
```diff
@@ -0,0 +1,9 @@
+<!-- wp:template-part {"slug":"header","theme":"emptytheme","tagName":"header"} /-->
```
I think we should not force a theme to specify the theme on each place the theme references a template part. The core should automatically assume that the template part being referenced is from that theme.

kjellr replied:
Yeah, definitely. This seems like an unnecessary attribute in this case.

jorgefilipecosta commented on emptytheme/block-templates/singular.html:
```diff
@@ -0,0 +1,5 @@
+<!-- wp:template-part {"slug":"header","theme":"emptytheme","tagName":"header"} /-->
```
The template parts are not loading on the edit-site screen but it is probably a bug with FSE.

kjellr replied:
I think that's likely due to [the Gutenberg change noted here](https://github.com/WordPress/gutenberg/issues/27072#issuecomment-733108038). I loaded this up on a fresh site just now and it loads as expected.

jorgefilipecosta reviewed (approved):
LGTM 👍

kjellr commented:
> `emptytheme/assets/block-template-parts` is an empty file that I think was accidentally created and checked in?

Good catch! That's removed now.

kjellr commented:
We just [discussed this PR in the Block-based Themes chat this week](https://wordpress.slack.com/archives/C02RP4Y3K/p1606922160403400), and decided that it makes sense to merge this PR in. It's been relatively solid the past couple weeks, and I think we hit all of the main discussion points in here. Plus, merging it will make @MaggieCabrera's theme generator easier for folks to use. 🙌

I'll go ahead and do that, and we can continue iterating in additional issues/PRs. Thanks everyone!
